### ISSUE TITLE
std::vector< std::vector< std::basic_string<char> > > not handled in Lib/r/std_vector.i

### ISSUE THREAD

**Issue by eirianop** [swig/swig#2385]
request to add support for a vector of vector of string for R please :)

**Comment by wsfulton** [comment]
How is it not handled? Have you tried this:

```swig
%include <std_string.i>
%include <std_vector.i>

%template(VectorString) std::vector<std::string>;
%template(VectorVectorString) std::vector<std::vector<std::string>>;
```

**Comment by eirianop** [comment]
Hi @wsfulton thank you for your prompt reply. Yes, I have tried these, and I notice the following behaviour when I try to call a function that returns std::vector<std::vector<std::string>> from R:

```
> h$vtest5()
An object of class "_p_std__vectorT_std__vectorT_std__string_std__allocatorT_std__string_t_t_std__allocatorT_std__vectorT_std__string_std__allocatorT_std__string_t_t_t_t"
Slot "ref":
<pointer: 0x55d64c224890>
```

As you can see, it is not being interpreted properly in R. As an example a std::vector<std::string>> is converted as expected:

```
> h$vtest3()
[1] "1" "2" "3" "4" "5"
```

I noticed that a vector of vectors is defined for many types in std_vector.i, but not for std::string and thought perhaps there was an omission. I have attempted to make code changes but unfortunately I don't know enough about the SWIG source code to complete this task at the present moment.

I have copied relevant contents from the interface, source, and header files below:

Hihi.i
```
%module Hihi

%{
  #include <string>
  #include "Hihi.hpp"
  #include "Utilities.hpp"
%}

%include "std_string.i"
%include "std_vector.i"
%include "stl.i"
namespace std {
  %template( vectori)   vector<int>;
  %template( vectors)   vector<string>;
  %template(vvectori)   vector< vector<int> >;
  %template(vvectors)   vector< vector<string> >;
};
%include "Hihi.hpp"
```

Hihi.hpp
```
#ifndef __HIHI_HPP__
#define __HIHI_HPP__

#include <vector>
#include <string>

class HihiReader {
    public:
        HihiReader(){};
        ~HihiReader(){};
        std::vector<std::string> vtest3();
        std::vector<std::vector<std::string> > vtest5();
        std::vector<std::vector<int> > vtest6();
};

#endif
```

Hihi.cpp
```
std::vector<std::string> HihiReader::vtest3() {
    std::vector<std::string> v = {"1", "2", "3", "4", "5"};
    return v;
}

std::vector<std::vector<std::string> > HihiReader::vtest5() {
    std::vector<std::vector<std::string> > vv;
    std::vector<std::string> v = {"1", "2", "3", "4", "5"};
    vv.push_back(v);
    return vv;
}

std::vector<std::vector<int> > HihiReader::vtest6() {
    std::vector<std::vector<int> > vv;
    std::vector<int> v = {1, 2, 3, 4, 5};
    vv.push_back(v);
    return vv;
}
```

and in the R shell
```
R version 3.4.4 (2018-03-15) -- "Someone to Lean On"
Copyright (C) 2018 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-gnu (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> dyn.load("libDataReader.so")
ource("Hihi.R")
cacheMetaData(1)> dyn.load("Hihi.so")
> source("Hihi.R")

> cacheMetaData(1)
> h <- HihiReader()
> h$vtest3()
[1] "1" "2" "3" "4" "5"
> h$vtest5()
An object of class "_p_std__vectorT_std__vectorT_std__string_std__allocatorT_std__string_t_t_std__allocatorT_std__vectorT_std__string_std__allocatorT_std__string_t_t_t_t"
Slot "ref":
<pointer: 0x55d64c224890>

> h$vtest3()
[1] "1" "2" "3" "4" "5"
> h$vtest6()
[[1]]
[1] 1 2 3 4 5
```

**Comment by wsfulton** [comment]
I morphed your code above into a SWIG example and used the `-debug-tmused` option to show the typemaps being used, comparing those returned by vtest3 with vtest5. 
```text
$ make -s clean check SWIGOPT="-debug-tmused"  | grep vtest[35]
example.h:11: Typemap for std::vector< std::string,std::allocator< std::string > > HihiReader::vtest3 (out) : %typemap(out) std::vector< std::string,std::allocator< std::string > >
example.h:11: Typemap for std::vector< std::string,std::allocator< std::string > > HihiReader::vtest3 (scoerceout) : %apply std::vector< std::basic_string< char > > { std::vector< std::string > }
example.h:11: Typemap for std::vector< std::string,std::allocator< std::string > > HihiReader::vtest3 (rtype) : %apply std::vector< std::basic_string< char > > { std::vector< std::string > }
example.h:12: Typemap for std::vector< std::vector< std::string,std::allocator< std::string > >,std::allocator< std::vector< std::string,std::allocator< std::string > > > > HihiReader::vtest5 (out) : %typemap(out) std::vector< std::vector< std::string,std::allocator< std::string > >,std::allocator< std::vector< std::string,std::allocator< std::string > > > >
example.h:12: Typemap for std::vector< std::vector< std::string,std::allocator< std::string > >,std::allocator< std::vector< std::string,std::allocator< std::string > > > > HihiReader::vtest5 (scoerceout) : %typemap(scoerceout) SWIGTYPE
example.h:12: Typemap for std::vector< std::vector< std::string,std::allocator< std::string > >,std::allocator< std::vector< std::string,std::allocator< std::string > > > > HihiReader::vtest5 (rtype) : %typemap(rtype) SWIGTYPE
```
You can see vtest5 is using SWIGTYPE typemaps instead of the custom ones which originate from https://github.com/swig/swig/blob/master/Lib/r/std_vector.i. You'll need copy and adapt the "scoerceout" and "rtype" typemaps from that file and include them into your own interface file. Hopefully not too hard!

**Comment by eirianop** [comment]
Hi @wsfulton could you point me to an example please?

For my own curiosity, could you explain why `std::vector<std::<vector<int> >` is defined in `std_vector.i` but `std::vector<std::<vector<std::string> >` isn't? 

Thanks again

**Comment by wsfulton** [comment]
Simply looks inconsistent and incomplete to me. I'd happy to accept a patch that adds in the missing vector<vector<X>> types, where X is std::basic_string<char> as well as the other missing types:, short, unsigned short, signed char, unsigned char. std_vector.i is the most relevant example I can see for this.

**Comment by eirianop** [comment]
Thanks for having a look @wsfulton. I have tried to add missing implementation but I do not feel that I have enough knowledge of SWIG internals to complete it. I'll be sure to create a pull request if this changes, however my lack of progress was the impetus to creating an issue 🙂

**Comment by wsfulton** [comment]
Fix will make it into the 4.1.0 release.

**Comment by eirianop** [comment]
That's awesome, thank you so much!
